patina_dxe_core/src/lib.rs

```diff
@@ -145,92 +236,148 @@ impl Default for GicBases {
 /// #     fn entry_point(self) -> patina::error::Result<()> { Ok(()) }
 /// # }
-/// # let physical_hob_list = core::ptr::null();
-/// patina_dxe_core::Core::default()
-///   .init_memory(physical_hob_list)
-///   .with_service(patina_ffs_extractors::CompositeSectionExtractor::default())
-///   .with_component(ExampleComponent::default())
-///   .start()
-///   .unwrap();
+/// struct ExamplePlatform;
+///
+/// impl ComponentInfo for ExamplePlatform {
+///   fn configs(mut add: Add<Config>) {
+///     add.config(32u32);
+///     add.config(true);
+///   }
+///
+///   fn components(mut add: Add<Component>) {
+///     add.component(ExampleComponent::default());
+///   }
+/// }
+///
+/// impl Platform for ExamplePlatform {
+///   type Extractor = NullSectionExtractor;
+///   type ComponentInfo = Self;
+///
+///   fn prioritize_32_bit_memory() -> bool {
+///     true
+///   }
+/// }
+///
+/// static CORE: Core<ExamplePlatform> = Core::new(NullSectionExtractor);
 /// ```
-pub struct Core<MemoryState> {
-    physical_hob_list: *const c_void,
-    hob_list: HobList<'static>,
-    components: Vec<Box<dyn Component>>,
-    storage: Storage,
-    platform_timer: PerfTimer,
-    _memory_state: core::marker::PhantomData<MemoryState>,
+pub struct Core<P: Platform> {
+    /// A parsed and heap-allocated list of HOBs provided by [Self::entry_point].
+    hob_list: OnceCell<HobList<'static>>,
+    /// The subsystem responsible for data management and dispatch of Patina components.
+    component_dispatcher: TplMutex<ComponentDispatcher>,
+    // NOTE: This field will be moved into the `DISPATCHER_CONTEXT` when it is moved into the Core.
+    section_extractor: P::Extractor,
 }
 
-impl Default for Core<NoAlloc> {
-    fn default() -> Self {
-        Core {
-            physical_hob_list: core::ptr::null(),
-            hob_list: HobList::default(),
-            components: Vec::new(),
-            storage: Storage::new(),
-            platform_timer: PerfTimer::new(),
-            _memory_state: core::marker::PhantomData,
-        }
+/// Safety: The Core struct can be safely accessed by multiple threads all internal mutability is protected by TPL
+/// locks for each subsystem.
+unsafe impl<P: Platform> Sync for Core<P> {}
+
+#[coverage(off)]
+impl<P: Platform> Core<P> {
+    /// Creates a new instance of the DXE Core in the NoAlloc phase.
+    pub const fn new(section_extractor: P::Extractor) -> Self {
+        Self { component_dispatcher: ComponentDispatcher::new_locked(), hob_list: OnceCell::new(), section_extractor }
     }
-}
 
-impl Core<NoAlloc> {
-    /// Initializes the timer frequency if a platform provides a custom implementation.
-    pub fn init_timer_frequency(self, frequency: Option<u64>) -> Core<NoAlloc> {
-        if let Some(freq_override) = frequency {
-            self.platform_timer.set_frequency(freq_override);
+    /// Sets the static DXE Core instance for global access.
+    #[allow(unused)]
+    fn set_instance(&'static self) -> bool {
```

Comment:
please create tests to verify set_instance set once behavior. That is

set -> succeeds
get -> returns set pointer
set -> panics

Reply:
panic is inside of one of the runtime functions so I'm not bothering with that, however I did add a test that ensures `set_instance` returns false on a second call (which leads to a panic in runtime), and that the address provided from `instance()` is the same address as the static core defined by the platform.